src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/Rfc3161TimestampRequest.cs

```diff
@@ -0,0 +1,432 @@
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.Pkcs
+{
+    public sealed class Rfc3161TimestampRequest
+    {
+        private byte[] _encodedBytes;
+        private Rfc3161TimeStampReq _parsedData;
+
+        private Rfc3161TimestampRequest()
+        {
+        }
+
+        public int Version => _parsedData.Version;
+        public ReadOnlyMemory<byte> GetMessageHash() => _parsedData.MessageImprint.HashedMessage;
+        public Oid HashAlgorithmId => _parsedData.MessageImprint.HashAlgorithm.Algorithm;
+        public Oid RequestedPolicyId => _parsedData.ReqPolicy;
+        public bool RequestSignerCertificate => _parsedData.CertReq;
+        public ReadOnlyMemory<byte>? GetNonce() => _parsedData.Nonce;
+        public bool HasExtensions => _parsedData.Extensions?.Length > 0;
+
+        public X509ExtensionCollection GetExtensions()
+        {
+            var coll = new X509ExtensionCollection();
+
+            if (!HasExtensions)
+            {
+                return coll;
+            }
+
+            X509ExtensionAsn[] rawExtensions = _parsedData.Extensions;
+
+            foreach (X509ExtensionAsn rawExtension in rawExtensions)
+            {
+                X509Extension extension = new X509Extension(
+                    rawExtension.ExtnId,
+                    rawExtension.ExtnValue.ToArray(),
+                    rawExtension.Critical);
+
+                // Currently there are no extensions defined.
+                // Should this dip into CryptoConfig or other extensible
+                // mechanisms for the CopyTo rich type uplift?
+                coll.Add(extension);
+            }
+
+            return coll;
+        }
+
+        public Rfc3161TimestampToken ProcessResponse(ReadOnlyMemory<byte> source, out int bytesConsumed)
+        {
+            Rfc3161RequestResponseStatus status;
+            Rfc3161TimestampToken token;
+
+            if (ProcessResponse(source, out token, out status, out int localBytesRead, shouldThrow: true))
+            {
+                Debug.Assert(status == Rfc3161RequestResponseStatus.Accepted);
+                bytesConsumed = localBytesRead;
+                return token;
+            }
+
+            Debug.Fail($"AcceptResponse should have thrown or returned true (status={status})");
+            throw new CryptographicException();
+        }
+
+        private bool ProcessResponse(
+            ReadOnlyMemory<byte> source,
+            out Rfc3161TimestampToken token,
+            out Rfc3161RequestResponseStatus status,
+            out int bytesConsumed,
+            bool shouldThrow)
+        {
+            status = Rfc3161RequestResponseStatus.Unknown;
+            token = null;
+
+            Rfc3161TimeStampResp resp;
+
+            try
+            {
+                resp = AsnSerializer.Deserialize<Rfc3161TimeStampResp>(source, AsnEncodingRules.DER, out bytesConsumed);
+            }
+            catch (CryptographicException)
```

Comment:
Nit: `catch (CryptographicException) when (!shouldThrow) { ... }` is generally preferred to rethrowing exceptions. This makes debugging a bit easier. See https://www.thomaslevesque.com/2015/06/21/exception-filters-in-c-6/ for more info.